core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+          if (oldRef == null || oldRef.get() != element)
+            ConcurrentWeakHashSet.UpdateResults.AddElement
+          else
+            ConcurrentWeakHashSet.UpdateResults.None
+      }
+    )
+  }
+
+  /**
+   * Remove given element from the set.
+   *
+   * @param element
+   *   element to remove
+   * @return
+   *   instance of the set
+   * @see
+   *   [[ConcurrentWeakHashSet.remove]]
+   */
+  override def subtractOne(element: V): ConcurrentWeakHashSet.this.type = {
+    this.remove(element)
+    this
+  }
+
+  /**
+   * Remove given element from the set.
+   *
+   * @param element
+   *   element to remove
+   * @return
+   *   `true` if the element was removed, `false` otherwise
+   */
+  override def remove(element: V): Boolean =
+    this.update(
+      element,
+      new ConcurrentWeakHashSet.UpdateOperation[V](
+        ConcurrentWeakHashSet.UpdateOptions.RestructureAfter,
+        ConcurrentWeakHashSet.UpdateOptions.SkipIfEmpty
+      ) {
+        override def execute(oldRef: Ref[V]): ConcurrentWeakHashSet.UpdateResults.UpdateResult =
+          if (oldRef != null && oldRef.get() == element)
+            ConcurrentWeakHashSet.UpdateResults.RemoveElement
+          else
+            ConcurrentWeakHashSet.UpdateResults.None
+      }
+    )
+
+  /**
+   * Remove all elements from the set.
+   */
+  override def clear(): Unit =
+    this.segments.foreach(segment => segment.clear())
+
+  /**
+   * Enhanced hashing same as in standard ConcurrentHashMap (Wang/Jenkins
+   * algorithm)
+   */
+  private def getHash(element: V): Int = {
+    var hash = if (element != null) element.hashCode() else 0
+    hash += (hash << 15) ^ 0xffffcd7d
+    hash ^= (hash >>> 10)
+    hash += (hash << 3)
+    hash ^= (hash >>> 6)
+    hash += (hash << 2) + (hash << 14)
+    hash ^= (hash >>> 16)
+    hash
+  }
+
+  /**
+   * Get segment for matched range of hashes.
+   */
+  private def getSegment(hash: Int): Segment =
+    this.segments((hash >>> (32 - this.shift)) & (this.segments.length - 1))
+
+  /**
+   * Get reference node for given element from matched segment.
+   * @see
+   *   [[ConcurrentWeakHashSet.getSegment]]
+   */
+  private def getReference(element: V): Ref[V] = {
+    val hash    = this.getHash(element)
+    val segment = this.getSegment(hash)
+    segment.getReference(element, hash)
+  }
+
+  /**
+   * Perform update operation on matched reference for provided element in the
+   * set.
+   */
+  private def update(element: V, update: ConcurrentWeakHashSet.UpdateOperation[V]): Boolean = {
+    val hash    = this.getHash(element)
+    val segment = this.getSegment(hash)
+    segment.update(hash, element, update)
+  }
+
+  /**
+   * Segment of the set that is responsible for a range of hashes. Locking
+   * mechanism is backed by ReentrantLock and dead references are cleaned up
+   * through ReferenceQueue.
+   */
+  private class Segment(initialSize: Int, var resizeThreshold: Int) extends ReentrantLock {
```

Comment:
Unfortunately, `ReentrantLock` is blocking. ZIO data structures are expected to be non-blocking. Although I realize, it's blocking on the level of a segment, which makes it faster than synchronizing the whole set, but nonetheless, it is still synchronizing.

I would consider using `AtomicReferenceArray`. Could that be sufficient to implement this logic?

Reply:
I was thinking about it, but the restructuring part of this implementation touches (writes to) several indices, so that exceeds the safe access to the given position guaranteed by AtomicReferenceArray. The lack of locking on the `references` (so effectively on the segment) could lead to the change of state between those updates and then to UB. Originally, ConcurrentHashMap used `ReentrantLock` as well, since JDK8 (stream api) they've moved back to regular `synchronized` closures on segments, but their impl is absolutely 6k+ long mess:

* https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java

so there's also an aspect of maintainability. At some point, when we need to restructure an array, there'll be always a locking point - it depends on if we're willing to trade performance/memory to reduce the impact of this.  

Just to be clear - I'm not aware of all pros/cons of `AtomicReferenceArray` here tho, but I assume they'd already use it if it'd be better than segments/nodes. I'll take a look through the current impl to see how it could work.